### DISCUSSION TITLE
Using StreamProvider with Sembast

### DISCUSSION THREAD

**Discussion by alyyasser** [rrousselGit/riverpod#451] [Q&A]
basically, i want to listen to record changes based on example from [change listener](https://github.com/tekartik/sembast.dart/blob/master/sembast/doc/change_listener.md). how i do it with StreamProvider in Riverpod? 

if i write something like this:
```
final appDataProvider = StreamProvider.autoDispose<AppData>((ref) async* {
  final db = await ref.watch(appDbProvider.future);
  final record = dataStore.record('app_data');
  final stream = record.onSnapshot(db);

  await for (final snapshot in stream) {
    if (snapshot != null) {
      yield AppData.fromJson(snapshot.value);
    }
  }
});
```
is that the right way? is there a better example on how should i do it?

**Comment by rrousselGit** [comment]
I don't understand your question. Is there a problem with the code you gave?

**Reply by alyyasser** [reply]
i haven't write the part which will use that provider, so im not sure if that will work or not.
im pretty much clueless about stream and still learning about it. could you tell me whether that code above is good enough, and im in the right direction?

**Reply by rrousselGit** [reply]
The code looks alright at first glance.

**Comment by simplenotezy** [comment]
@alyyasser what about this?

```dart
  @override
  Stream<Cart> watchCart() {
    final record = store.record(cartItemsKey);
    return record.onSnapshot(db).map((snapshot) {
      if (snapshot != null) {
        return Cart.fromJson(snapshot.value);
      } else {
        return const Cart();
      }
    });
  }
```

```dart
final cartProvider = StreamProvider<Cart>((ref) {
  return ref.watch(localCartRepositoryProvider).watchCart();
});
```

**Reply by alyyasser** [reply]
yours definitely works and looks cleaner too (with correct layer separation). though, my original code works fine and i still using it right now. i might write something like yours when i refractor my codebase later.
